### ISSUE TITLE
Adding Manual Buttons

### ISSUE THREAD

**Issue by Jambuster1** [Aircoookie/Espalexa#42]
Hello,

Currently have your code implemented to control the opening of some blinds. However, I want a way to add a button in so I can change some sort of 'Mode' so that I can then use another two buttons for manual control. Then when the button is pressed again, it goes back to using Alexa. I was just wondering if this is possible?

Even better would be to eliminate the mode buttons and just have two buttons which can directly override anything. If you could help that would be great!

Many Thanks
Jamie

**Comment by Aircoookie** [comment]
Hi Jamie,
sure, that is possible! I don't know exactly how your sketch controls the blinds, so I am assuming you use some relais connected to digital pins for it. You can add a function to the loop() in your sketch to check the status of your up and down buttons and write to the relais just like you do in the lightChanged() Espalexa callback. If you need help, there are a lot of button tutorials out there and this helpful library: [JC_Button](https://github.com/JChristensen/JC_Button).

You can add any functionality you want to your sketch, Espalexa is just a library that enables Alexa support, the ESP is capable of doing a lot more stuff "simultaneously" :)

**Comment by Jambuster1** [comment]
Hi Aircoookie

Sorry, I should of Clarified how I’m controlling it. I’m using a standard servo which is being controlled using your basic code example but instead of controlling lights, I’m setting the percentage and this in turn relates to the angle I want the servo at.

The issue is that I tried creating code in the loop section which works. However, as soon as I enter commands through Alexa, I can no longer use the buttons.

It seems to be that as soon as the espalexa.loop() runs, the buttons no longer run? 

Jamie

**Comment by Aircoookie** [comment]
Ah ok, now I understand. Your button functions in the loop() should still work though. Can you post your code so I can take a look at what's wrong?

**Comment by Jambuster1** [comment]
I will have to send you the code later. I’m currently away from my PC but will send it as soon as I’m back.

Thanks
Jamie

**Comment by Jambuster1** [comment]
Sorry for the delay in getting back to you. Please find below the code I tried to use with the two buttons:

`
        #include <ESP8266WiFi.h>
#include <Espalexa.h>
#include <Servo.h>

Servo blindServo;

static const int servoPin = 15;
const int button1Pin = 5;
const int button2Pin = 4;

int servoPosition = 0;

// prototypes
boolean connectWifi();

//callback functions
void servoPositionChanged(uint8_t servoPosition);

// Change this!!
const char* ssid = "TP-LINK_2.4GHz";
const char* password = "Khag7IffA7es";

boolean wifiConnected = false;

Espalexa espalexa;

void setup()
{
  Serial.begin(115200);

  blindServo.attach(servoPin);

  pinMode(button1Pin, INPUT);
  pinMode(button2Pin, INPUT);
  
  // Initialise wifi connection
  wifiConnected = connectWifi();

  if (wifiConnected) {

    // Define your devices here.
    espalexa.addDevice("Blinds", servoPositionChanged, 20); //simplest definition, default state off
    espalexa.begin();

  } else
  {
    while (1) {
      Serial.println("Cannot connect to WiFi. Please check data and reset the ESP.");
      delay(2500);
    }
  }
}


void loop()
{
  while (digitalRead(button1Pin) == HIGH && servoPosition < 179) {
    servoPosition++;
    blindServo.write(servoPosition);
    delay(15);
  }
  while (digitalRead(button2Pin) == HIGH && servoPosition > 0) {
    servoPosition--;
    blindServo.write(servoPosition);
    delay(15);
  }


  
  espalexa.loop();
  delay(1);
}


//our callback functions
void servoPositionChanged(uint8_t servoPosition) {
  Serial.print("Servo Position changed to ");

  //EXAMPLE
  if (servoPosition < 180) {

    blindServo.attach(servoPin);

    Serial.print("Servo Position ");
    Serial.println(servoPosition);

    blindServo.write(servoPosition);

    delay(1000);

    blindServo.detach();

  }
  else  {
    Serial.print("Servo Out of Range ");
    Serial.println(servoPosition);
  }
}

// connect to wifi – returns true if successful or false if not
boolean connectWifi() {
  boolean state = true;
  int i = 0;

  WiFi.mode(WIFI_STA);
  WiFi.begin(ssid, password);
  Serial.println("");
  Serial.println("Connecting to WiFi");

  // Wait for connection
  Serial.print("Connecting...");
  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
    Serial.print(".");
    if (i > 20) {
      state = false; break;
    }
    i++;
  }
  Serial.println("");
  if (state) {
    Serial.print("Connected to ");
    Serial.println(ssid);
    Serial.print("IP address: ");
    Serial.println(WiFi.localIP());
  }
  else {
    Serial.println("Connection failed.");
  }
  return state;
}

`

**Comment by Aircoookie** [comment]
You have `blindServo.detach();` in the alexa callback function. This disables the servo so that the buttons in the loop() no longer have any effect. Please remove the attach and detach commands from the callback function. If that doesn't work (there could be some issues with using servo PWM and WiFi simultaneously) add the attach and detach inside the button while loops as well.

That should fix your issue :)

**Comment by Jambuster1** [comment]
Ah yes of course!! I will try adding attach and detach in the while loop as I need the servo to detach. Another quick question which you may be able to answer is, whats the best sleep mode to put the ESP8266 into for this code to still work? I need to use as little power as possible?

Jamie

**Comment by Aircoookie** [comment]
The default sleep mode is modem sleep, where the WiFi modem is turned off while there is no data transmitted. It uses ~22mA on average. You can try if light sleep reduces your current usage (it might cause issues though) by adding `wifi_set_sleep_type(LIGHT_SLEEP_T);` at the end of `setup()` (in the `if (wifiConnected)` block). Also increase the delay in loop after `espalexa.loop()` to `delay(250);`. If you are lucky, this will reduce power draw to 10mA, if you are unlucky, it will no longer work and modem sleep is the best solution that works. Deep sleep would be best and uses <1mA, but is not an option since the ESP must be able to receive Alexa commands all the time. Deep sleep can only be used for sensors where a permanent connection is not required.
